    public void initEnglishDoc() {
        helpDoc = "<html><body><p><b><font size=5>Introduction</font></b></p><p><font size=3>Thank you for choosing MSA2SNP, which provides data mining and visualizing from multiple sequence alignments (MSA). You can get the latest version from <a href=\"http://sourceforge.net/projects/msa2snp\">http://sourceforge.net/projects/msa2snp</a>. </font></p><p><b><font size=5>System Requirements</font></b></p><p><font size=3>MSA2SNP runs on Microsoft Windows 98/ME/2000/XP/Vista, Linux, Solaris or MacOSX. Using MSA2SNP requires Java installation. You can get the latest stable release of Java Runtime Environment from <a href=\"http://java.sun.com\">Sun Microsystems</a>. The recommended system specification is listed as below:</font></p><p><font size=3><b>CPU</b>: Pentium III or higher</font></p><p><font size=3><b>RAM</b>: 512M or higher</font></p><p><font size=3><b>DISPLAY</b>: 1024x768 resolution, 256 colors&nbsp; or higher. </font></p><p><font size=3><b>OTHER REQUIREMENTS</b>: Mouse or tablet</font></p><p><b><font size=5>Installation</font></b></p><p><font size=3>MSA2SNP is provided as a runnable jar file. To start MSA2SNP, type &quot;java -jar MSA2SNP.jar&quot; in your console or terminal in your are using Linux. For Windows users, just double-click MSA2SNP.jar. </font></p><p><font size=5><b><span lang=\"zh-cn\">Feature</span>s</b></font></p><p><font size=3>- Browse multiple sequence alignment files (*.aln). </font></p><p><font size=3>- View mutation sites, types and their frequencies. </font></p><p><font size=3>- Highlight constant sites, variable sites, parsimony-informative sites and singleton sites. </font></p><p><font size=3>- Filter by mutation frequencies, distances or qualities. </font></p><p><font size=3>- View statistics of sequences or mutation sites. </font></p><p><font size=5><b>Menus</b></font></p><p><font size=4><b>1. File</b></font></p><p><font size=3>1.1 <b>Open</b>: Use this command to import a file containing sequences (*.fas, *.aln). The qualities of bases can also be imported if you prefer. The format of base quality file is like this: </font></p><p><font size=3>&gt;sequence1</font></p><p><font size=3>99 98 12 99 88 ...</font></p><p><font size=3>&gt;sequence2</font></p><p><font size=3>98 88 12 99 98 ...</font></p><p><font size=3>1.2 <b>Exit</b>: Use this command to quit the program.</font></p><p><font size=4><b>2. </b></font><b><font size=4>View</font></b></p><p><font size=3>The items below this menu represent the </font>alignment files you have already opened. Click items to jump among those active sessions. </p><p><font size=4><b>3. </b></font><b><font size=4>Tools</font></b></p><p>3.1 <b>Language</b>: Use <font size=3>this command to choose a language of the GUI. The default language is English. Simplified Chinese is available.</font></p><p><font size=3>3.2 <b>Options</b>: Use this command to set path for <b>clustalw</b> program and path to place temporary files.&nbsp; </font></p><p><font size=4><b>4. Run</b></font></p><p>4.1 <b>Run ClustalW</b>: Use <font size=3>this command to perform a</font><b>clustalw</b> alignment of input sequences. Once the alignment process is completed, the results are loaded automatically and ready for your operations. </p><p><b><font size=4>5. Help</font></b></p><p><font size=3>5.1 <b>Help</b>: </font>Use <font size=3>this command to show help topics.</font></p><p><font size=3>5.2 <b>About</b>: </font>Use <font size=3>this command to display the copyright notice and version of your copy of MSA2SNP</font></p><p><b><font size=5>Toolbar</font></b></p><p><font size=3>The toolbar contains four buttons (C, V, Pi, S) and three drop lists (Filter by frequency, Filter by distance, Filter by quality). </font></p><p><font size=4><b>1. Highlighting Sites</b></font></p><p><font size=3><b>C</b>: If this button is pressed, then all constant sites (contains at least two sequences contain unambiguous nucleotides or amino acids) will be highlighted. A count of the highlighted sites will be displayed on the status bar.<br><br><b>V</b>: If this button is pressed, then all variable sites (contains at least two types of nucleotides or amino acids) will be highlighted. A count of the highlighted sites will be displayed on the status bar.<br><br><b>Pi</b>: If this button is pressed, then all parsimony-informative sites (contains at least two types of nucleotides or amino acids, and at least two of them occur with a minimum frequency of two) will be highlighted. A count of the highlighted sites will be displayed on the status bar.<br><br><b>S</b>: If this button is pressed, then all singleton sites (contains at least two types of nucleotides or amino acids with, at most, one occurring multiple times) will be highlighted. A count of the highlighted sites will be displayed on the status bar.</font></p><p><font size=4><b>2. Filter by Conditions</b></font></p><p><font size=3><b>Filter by frequency</b>: Set the color and font for the nucleotides or amino acids which you wish to be highlighted within the range of the minimum and maximum frequencies. </font></p><p><font size=3><b>Filter by distance</b>: Set the color and font for the nucleotides or amino acids which you wish to be highlighted which is greater than the minimum distance. </font></p><p><font size=3><b>Filter by quality</b>: Set the color and font for the nucleotides or amino acids which you wish to be highlighted which is greater than the minimum quality. </font></p><p><font size=4><b>&nbsp;3. Show Statistics Information</b></font></p><p><font size=3>The panel containing statistics information is below the sequence table, which can be toggle display or hide with button. While statistics panel is shown, if you click a nucleotide or amino acid in the table, statistics of the sequence which contains that base and the mutation frequency of that site are shown in the panel. </font></p><p><b><font size=5>Copyright</font></b></p><p><font size=3>THE PROGRAM IS PROVIDED &quot;AS IS&quot; WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING, BUT NOT LIMITED TO WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT WILL THE AUTHOR or AUTHORS BE LIABLE TO YOU FOR ANY DAMAGES, INCLUDING INCIDENTAL OR CONSEQUENTIAL DAMAGES, ARISING OUT OF THE USE OF THE PROGRAM, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.<br><br>YOU ACKNOWLEDGE THAT YOU HAVE READ THIS LICENSE, UNDERSTAND IT AND AGREE TO BE BOUND BY ITS TERMS AS THE COMPLETE AND EXCLUSIVE STATEMENT OF THE AGREEMENT BETWEEN US, SUPERSEDING ANY PROPOSAL OR PRIOR AGREEMENT, ORAL OR WRITTEN, AND ANY OTHER COMMUNICATIONS BETWEEN US RELATING TO THE SUBJECT MATTER OF THIS LICENSE.</font></p><p><font size=3>MSA2SNP is free software; you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation; either version 2.1 of the License, or (at your option) any later version.. You should have received a copy of the GNU Lesser General Public License, </font></p><p>The program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.<br><br>You should have received a copy of the GNU Lesser General Public License along with this library; if not, you can get a copy from<a href=\"http://www.gnu.org/licenses/lgpl.html\">http://www.gnu.org/licenses/lgpl.html</a> or write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA</p><p><b><font size=5>Contact</font></b></p><p>For more questions or bug report, please e-mail to<a href=\"mailto:qwang.big@gmail.com\">qwang.big@gmail.com</a> or post to open discussion forum<a href=\"http://sourceforge.net/forum/forum.php?forum_id=913637\">http://sourceforge.net/forum/forum.php?forum_id=913637</a>. </p></body></html>";
    }
